fix: force Jackson transitive dependencies versions

loicmathieu opened the pull request:
Force the version of all the Jackson dependencies that are inherited from smallrye-openapi.

I cannot find an other way than duplicating the jackson.version property as we need to use the dependencyManagement part of the pom. I add a comment to the Quarkus bom to avoid missing an update of the property in the two places.

Fixes #2789

loicmathieu commented:
With these changes, the dependency tree shows that the Jackson version are correct for all the Jackson modules.

```
] --- maven-dependency-plugin:3.1.1:tree (default-cli) @ quarkus-smallrye-openapi ---
[INFO] io.quarkus:quarkus-smallrye-openapi:jar:999-SNAPSHOT
[INFO] +- io.smallrye:smallrye-open-api:jar:1.1.3:compile
[INFO] |  +- org.eclipse.microprofile.openapi:microprofile-openapi-api:jar:1.1.2:compile
[INFO] |  |  \- org.osgi:org.osgi.annotation.versioning:jar:1.0.0:compile
[INFO] |  +- org.eclipse.microprofile.config:microprofile-config-api:jar:1.3:compile
[INFO] |  +- com.fasterxml.jackson.core:jackson-core:jar:2.9.9:compile
[INFO] |  +- com.fasterxml.jackson.core:jackson-databind:jar:2.9.9:compile
[INFO] |  |  \- com.fasterxml.jackson.core:jackson-annotations:jar:2.9.9:compile
[INFO] |  +- com.fasterxml.jackson.dataformat:jackson-dataformat-yaml:jar:2.9.9:compile
[INFO] |  |  \- org.yaml:snakeyaml:jar:1.23:compile
[INFO] |  +- org.jboss:jandex:jar:2.1.1.Final:compile
[INFO] |  +- org.jboss.logging:jboss-logging:jar:3.3.2.Final:compile
[INFO] |  +- org.jboss.shrinkwrap:shrinkwrap-api:jar:1.2.6:compile
[INFO] |  \- commons-beanutils:commons-beanutils:jar:1.9.3:compile
[INFO] |     +- commons-logging:commons-logging:jar:1.2:compile
[INFO] |     \- commons-collections:commons-collections:jar:3.2.2:compile
[INFO] +- io.quarkus:quarkus-core:jar:999-SNAPSHOT:compile
[INFO] |  +- javax.inject:javax.inject:jar:1:compile
[INFO] |  +- javax.annotation:javax.annotation-api:jar:1.3.2:compile
[INFO] |  +- io.smallrye:smallrye-config:jar:1.3.5:compile
[INFO] |  |  \- javax.enterprise:cdi-api:jar:2.0.SP1:compile
[INFO] |  |     +- javax.el:javax.el-api:jar:3.0.0:compile
[INFO] |  |     \- javax.interceptor:javax.interceptor-api:jar:1.2:compile
[INFO] |  +- org.jboss.logmanager:jboss-logmanager-embedded:jar:1.0.3:compile
[INFO] |  +- org.jboss.logging:jboss-logging-annotations:jar:2.1.0.Final:compile
[INFO] |  +- org.jboss.threads:jboss-threads:jar:3.0.0.Beta4:compile
[INFO] |  +- org.slf4j:slf4j-api:jar:1.7.25:compile
[INFO] |  +- org.jboss.slf4j:slf4j-jboss-logging:jar:1.1.0.Final:compile
[INFO] |  +- org.graalvm.sdk:graal-sdk:jar:1.0.0-rc16:compile
[INFO] |  +- org.wildfly.common:wildfly-common:jar:1.5.0.Final-format-001:compile
[INFO] |  +- org.wildfly.security:wildfly-elytron-ssl:jar:2.0.0.Alpha4:compile
[INFO] |  |  +- org.wildfly.security:wildfly-elytron-auth-server:jar:2.0.0.Alpha4:compile
[INFO] |  |  |  \- org.wildfly.security:wildfly-elytron-permission:jar:2.0.0.Alpha4:compile
[INFO] |  |  +- org.wildfly.security:wildfly-elytron-base:jar:2.0.0.Alpha4:compile
[INFO] |  |  +- org.wildfly.security:wildfly-elytron-credential:jar:2.0.0.Alpha4:compile
[INFO] |  |  |  +- org.wildfly.security:wildfly-elytron-keystore:jar:2.0.0.Alpha4:compile
[INFO] |  |  |  \- org.wildfly.security:wildfly-elytron-provider-util:jar:2.0.0.Alpha4:compile
[INFO] |  |  +- org.wildfly.security:wildfly-elytron-auth:jar:2.0.0.Alpha4:compile
[INFO] |  |  +- org.wildfly.security:wildfly-elytron-util:jar:2.0.0.Alpha4:compile
[INFO] |  |  \- org.wildfly.security:wildfly-elytron-x500:jar:2.0.0.Alpha4:compile
[INFO] |  \- org.wildfly.security:wildfly-elytron-x500-cert:jar:2.0.0.Alpha4:compile
[INFO] |     +- org.wildfly.security:wildfly-elytron-asn1:jar:2.0.0.Alpha4:compile
[INFO] |     \- org.wildfly.security:wildfly-elytron-x500-cert-util:jar:2.0.0.Alpha4:compile
[INFO] +- io.quarkus:quarkus-undertow:jar:999-SNAPSHOT:compile
[INFO] |  +- io.undertow:undertow-servlet:jar:2.0.19.Final:compile
[INFO] |  |  +- io.undertow:undertow-core:jar:2.0.19.Final:compile
[INFO] |  |  \- org.jboss.spec.javax.servlet:jboss-servlet-api_4.0_spec:jar:1.0.0.Final:compile
[INFO] |  +- org.jboss.xnio:xnio-nio:jar:3.7.1.Final:compile
[INFO] |  \- org.jboss.xnio:xnio-api:jar:3.7.1.Final:compile
[INFO] +- io.quarkus:quarkus-arc:jar:999-SNAPSHOT:compile
[INFO] |  +- io.quarkus.arc:arc:jar:999-SNAPSHOT:compile
[INFO] |  \- org.eclipse.microprofile.context-propagation:microprofile-context-propagation-api:jar:1.0-RC1:compile
[INFO] \- io.quarkus:quarkus-swagger-ui:jar:999-SNAPSHOT:compile
```

kenfinnigan commented:
Is there no way to see where it's being overridden from?

gsmet reviewed (changes requested):
See my comment here https://github.com/quarkusio/quarkus/issues/2789#issuecomment-500841211 about potentially importing the bom. Please do not rush this one until we get this thing settled.

loicmathieu commented:
@kenfinnigan the issue comes from the pom.xml of io.smallrye:smallrye-open-api:jar:1.1.3 that references Jackson 2.9.8 and as no extentions import the Quarkus bom (that references 2.9.9) we need to fix the version on the quarkus-smallrye-openapi pom too.

Or we can import the Quarkus bom inside the extention but I didn't follow this way as no extention import the Quarkus bom ...

Importing the Jackson bom into the extention will not solves the duplicate version issue as we still needs to duplicate the version property in both the Quarkus bom and the extention pom but it will avoid to list the Jackson modules that smallrye-openapi depends on so maybe it's a better way to do it, I can change my PR.

kenfinnigan commented:
We can't add more exclusions to `smallrye-open-api` dependency?

loicmathieu commented:
I just tested importing the Quarkus bom in the extention and it's not solving the issue.

So we need to duplicate the Jackson version in the smallrye-openapi extention and add the Jackson jar versions in the dependencyManagement section of the pom.

For the dependencyManagement section, my current PM list the JARs that are used by the extention, replacing them with the Jackson bom also works so I can change my PR to do so if needed.

I already tested a lot of different approach without much success ...

gsmet commented:
@loicmathieu the Jackson bom would need to be included in the Quarkus bom with the other boms already there.

loicmathieu commented:
For the sake of clarity I resume here all the stuff around this issue:
- The Jackson bom **is** imported in the Quarkus bom
- The Quarkus bom **is not imported** in the extention bom but importing didn't solve the issue
- We **must** fix the version on the dependencyManagement section of the extention (I tested a few other ways with no success).
- I don't find a better way than duplicated the jackson.version property inside the extention pom
- There is two ways to fix the issue, both on the extention pom, the one provided by this PR and a slightly different version that set the version on the dependencyManagement section for all Jackson jars using the bom instead of listing them as I did, I can update my PR accordingly

I know that this is not optimal, but I tested _a lot_ of other way to do it and none works ...

loicmathieu committed:
fix(#2789): fix Jackson version issue

The oder of import in the bom is important, we need to import Jackson first before RESTEasy.
After this, exluding and importing Jackson in the quarkus-smallrye-openapi extention is useless.

loicmathieu commented:
Thanks to @aloubyansky there is a better way by inversing the import order on the Quakrus bom as order is important and RESTEasy also import Jackson.
There is still an issue with the jackson-annotation dependency that now comes in 2.9.0, I'll have a look at it. I just launch a dependency check on all Quarkus and will investiguate ...

aloubyansky commented:
The annotation artifact is set to that version in the jackson-bom.

loicmathieu commented:
@aloubyansky I update my PR to include a specific dependencyManagement entry for the jackson-annotation module. It's needed for some obscure dependency management issue with resteasy.

With it, everything is fine, without it this specific jar only comes in 2.9.0 whereas the others comes in 2.9.9.

This PR is ready to GO !

gsmet commented:
@loicmathieu AFAICS, the `jackson-bom` mandates `jackson-annotations` `2.9.0`. That's why it's the version used.

I think it's probably because they don't change the annotations in micro versions. In any case, I would let it like that in Quarkus (e.g. don't force it and rely on the bom) and ask the question to the Jackson developers.

loicmathieu commented:
@gsmet in the current situation, the version of the annotation module is 2.9.8 (see the dependency tree output on the issue). So if I remove the dependencyManagement addon it will fall back to a different version.
But as it I understand it, there is no difference between all the patch version of the annotation module, i the 3.x series, there will be not patch level dependencies anymore ... but for Java 9 compatibility we should at least use the 2.9.1 version as it has an automatic module name. As Quarkus is compatible with  Java 11 maybe we should use this version and not the 2.9.0.

gsmet commented:
@loicmathieu OK, let's force it then. I think it would be worth opening a Jackson issue to see what they have to say about this. It seems to be a conscious choice of them as it's explicitly set in the bom but it looks a bit weird.

gsmet commented on bom/pom.xml:
```diff
@@ -195,6 +186,23 @@
                 <scope>import</scope>
                 <type>pom</type>
             </dependency>
+            <!-- We need to fix this specific Jackson module version due to dependency management issue -->
```
I think this comment should be updated with something like:

> The Jackson BOM imports version 2.9.0 and we want all the versions to be in sync. Especially, 2.9.1 adds an automatic module name.

loicmathieu replied:
done

gsmet commented on bom/pom.xml:
```diff
@@ -195,6 +186,23 @@
+
+
+            <!-- RESTEasy dependencies, imported as a BOM -->
+            <!-- As RESTEasy also import Jackson, we need to always import it after Jackson -->
```
```suggestion
            <!-- As RESTEasy also imports Jackson, we need to import its BOM after the Jackson BOM -->
```

loicmathieu replied:
done

gsmet reviewed (changes requested):
@loicmathieu I think we're nearly there. Thanks for tackling that one.

Could you integrate my suggestion if they make sense to you? (no need for a separate commit, you can squash them in your commit)

gsmet reviewed (approved):
Let's wait for CI.

FroMage commented:
CI is green, shall I merge it @gsmet ?

gsmet commented:
Done! Thanks @loicmathieu and @aloubyansky
